Marlin/src/gcode/config/M305.cpp

```diff
@@ -0,0 +1,137 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "../../inc/MarlinConfig.h"
+
+#if ENABLED(USES_THERMISTOR_PARAMS)
+
+#include "../gcode.h"
+#include "../../module/temperature.h"
+
+inline void log_user_thermistor(PGM_P const prefix, const user_thermistor_t &t) {
+  if (prefix != NULL) {
+    SERIAL_ECHO_START();
+    serialprintPGM(prefix);
+  }
+  SERIAL_PRINTF(" r:%0.1f", t.series_res);
+  SERIAL_PRINTF(" t:%0.1f", t.res_25);
+  SERIAL_PRINTF(" b:%0.1f", t.beta);
+  SERIAL_PRINTF(" c:%0.9f", t.sh_c_coeff);
+  SERIAL_EOL();
+}
+
+void GcodeSuite::M305() {
+  constexpr int num_thermistors = COUNT(thermalManager.user_thermistor);
+  int t_index = -1;
+
+  if (!parser.seenval('P')) {
+    //SERIAL_ECHO_MSG("   format: M305 P<thermistor_number> R<pull-up_resistor_value> T<thermistor_resistance_at_25C> B<thermistor_beta> C<Steinhart-Hart C coefficient>");
+    //SERIAL_ECHO_MSG(" examples: M305 P0 R4700 T100000 B3950 C0.0");
+    //SERIAL_ECHO_MSG("           M305 P0 R4700");
+    //SERIAL_ECHO_MSG("           M305 P0 T100000");
+    //SERIAL_ECHO_MSG("           M305 P0 B3950");
+    //SERIAL_ECHO_MSG("           M305 P0 C0.0");
+  }
+  else {
+    t_index = parser.value_int();
+    if (!WITHIN(t_index, 0, num_thermistors - 1)) {
+      SERIAL_ECHO_MSG(" error: Invalid thermistor index. Must be 0 to ");
+      SERIAL_ECHOLN(num_thermistors - 1);
+      return;
+    }
+
+    user_thermistor_t &t = thermalManager.user_thermistor[t_index];
+
+    if (parser.seen('R')) {  // pull-up resistor value
+      const float res = parser.value_float();
+      if (res <= 0 || res >= 1000000)
+        SERIAL_ECHO_MSG(" error: Invalid series resistance. Must be > 0 and < 1000000");
+      else
+        t.series_res = res;
+    }
+
+    if (parser.seen('T')) {  // thermistor resistance at 25C
+      float res = parser.value_float();
+      if (res <= 0 || res >= 10000000)
+        SERIAL_ECHO_MSG(" error: Invalid thermistor resistance @ 25C. Must be > 0 and < 10000000");
+      else {
+        t.res_25 = res;
+        t.pre_calc = true;
+      }
+    }
+
+    if (parser.seen('B')) { // thermistor beta value
+      float beta = parser.value_float();
+      if (beta <= 0 || beta >= 1000000)
+        SERIAL_ECHO_MSG(" error: Invalid thermistor beta. Must be > 0 and < 1000000");
+      else {
+        t.beta = beta;
+        t.pre_calc = true;
+      }
+    }
+
+    if (parser.seen('C')) { // thermistor Steinhart-Hart C coefficient
+      float coeff = parser.value_float();
+      if (coeff <= -0.01f || coeff >= 0.01f)
+        SERIAL_ECHO_MSG(" error: Invalid thermistor Steinhart-Hart C coefficient. Must be > -0.01 and < +0.01");
+      else {
+        t.sh_c_coeff = coeff;
+        t.pre_calc = true;
+      }
+    }
+  }
+
```

Comment:
I will continue to compress these messages and adjust them to fit conventions. The word "error" is problematic, as it can trigger some hosts to disconnect, but no worries — there is an established alternative.

Reply:
no problem.

I guess you want to add this addition to Marlin then ? .. cool

Reply:
So far it is intriguing me. I see it has gotten a lot of interest and feedback too.